### ISSUE TITLE
Overloading resolution appears to fail when using more than one type class

### ISSUE THREAD

**Issue by kierdavis** [nim-lang/Nim#4745]
``` nim
proc test1*[T1: uint|uint8|uint16|uint32|uint64, T2: SomeInteger](x: T1, y: T2) =
  discard
proc test2*[T1: SomeUnsignedInt, T2: uint|uint8|uint16|uint32|uint64|int|int8|int16|int32|int64](x: T1, y: T2) =
  discard
proc test3*[T1: SomeUnsignedInt, T2: SomeInteger](x: T1, y: T2) =
  discard

test1(1u32, 5u8) # works
test2(1u32, 5u8) # works
test3(1u32, 5u8) # Error: type mismatch: got (uint32, uint8)
```

Probably more likely than not that I'm doing something wrong here, but if this is a compiler bug then it looks like it could be related to #4375.

**Comment by kierdavis** [comment]
Looks like this is intended behaviour. When trying to match `test3(1u32, 5u8)`, the first argument is checked against the type class `SomeUnsignedInt`, which matches but also instantiates `SomeUnsignedInt` to type `uint32`. Then, when checking the second argument, `SomeInteger` effectively expands to `SomeSignedInt|uint32` rather than `SomeSignedInt|SomeUnsignedInt`, and so `uint8` will fail to match.

According to the manual,

> By default, during overload resolution each named type class will bind to exactly one concrete type.

so this is the intended behaviour. The manual goes on to say:

> Alternatively, the `distinct` type modifier can be applied to the type class to allow each param matching the type class to bind to a different type.

though I can't seem to get this to work; none of the following work:

```nim
proc test3*[T1: distinct SomeUnsignedInt, T2: SomeInteger](x: T1, y: T2)
proc test3*[T1: distinct SomeUnsignedInt, T2: distinct SomeInteger](x: T1, y: T2)
proc test3*[T1: distinct SomeUnsignedInt, T2: SomeSignedInt | distinct SomeUnsignedInt](x: T1, y: T2)
```

**Comment by Araq** [comment]
> Then, when checking the second argument, SomeInteger effectively expands to SomeSignedInt|uint32 rather than SomeSignedInt|SomeUnsignedInt, and so uint8 will fail to match.

No, there is no such connection between ``T1`` and ``T2`` and it's a horrible compiler bug.

**Comment by dom96** [comment]
This code now compiles for me.

**Comment by markus-oberhumer** [comment]
Just for the record this bug seems fixed in nim 0.16.0 (it fails with 0.15.2).

I cannot find any references in the git repo, though.
